File: examples/cranedns-api.py
#!/usr/bin/env python3
"""
CraneDNS API Client Example - Python

This script demonstrates how to use all CraneDNS API methods.

Requirements:
    pip install requests

Usage:
    1. Set your API_KEY and BASE_URL below
    2. Run: python cranedns-api.py
"""

import requests
import json

# Configuration
API_KEY = "your-api-key"
BASE_URL = "https://namecrane.org/index.php"

headers = {
    "X-API-Key": API_KEY,
    "Content-Type": "application/json"
}


def list_records(record_type=None):
    """
    List all DNS records, optionally filtered by type.

    Example Response:
    {
        "success": true,
        "records": [
            {"name": "@", "type": "A", "content": "192.168.1.1", "ttl": 3600},
            {"name": "www", "type": "CNAME", "content": "example.com.", "ttl": 3600}
        ]
    }
    """
    params = {"m": "cranedns", "action": "api", "method": "list"}
    if record_type:
        params["type"] = record_type

    response = requests.get(BASE_URL, params=params, headers=headers)
    return response.json()


def create_record(name, record_type, content, ttl=3600):
    """
    Create a new DNS record.

    Example Response:
    {"success": true, "message": "Record created"}

    Error Response:
    {"success": false, "error": "Invalid IPv4 address format"}
    """
    response = requests.post(
        BASE_URL,
        params={"m": "cranedns", "action": "api", "method": "create"},
        headers=headers,
        json={
            "name": name,
            "type": record_type,
            "content": content,
            "ttl": ttl
        }
    )
    return response.json()


def update_record(name, record_type, old_content, new_content, ttl=3600):
    """
    Update an existing DNS record.

    Example Response:
    {"success": true, "message": "Record updated"}

    Error Response:
    {"success": false, "error": "name, type, old_content, and content are required"}
    """
    response = requests.post(
        BASE_URL,
        params={"m": "cranedns", "action": "api", "method": "update"},
        headers=headers,
        json={
            "name": name,
            "type": record_type,
            "old_content": old_content,
            "content": new_content,
            "ttl": ttl
        }
    )
    return response.json()


def delete_record(name, record_type, content):
    """
    Delete a DNS record.

    Example Response:
    {"success": true, "message": "Record deleted"}

    Error Response:
    {"success": false, "error": "name, type, and content are required"}
    """
    response = requests.post(
        BASE_URL,
        params={"m": "cranedns", "action": "api", "method": "delete"},
        headers=headers,
        json={
            "name": name,
            "type": record_type,
            "content": content
        }
    )
    return response.json()


def ddns_update(name, record_type="A"):
    """
    Update A/AAAA record with the client's current IP address.

    Example Response:
    {
        "success": true,
        "message": "DDNS updated",
        "name": "home",
        "type": "A",
        "content": "203.0.113.45"
    }

    Error Response:
    {"success": false, "error": "Could not detect a valid IPv6 address"}
    """
    response = requests.get(
        BASE_URL,
        params={
            "m": "cranedns",
            "action": "api",
            "method": "ddns",
            "name": name,
            "type": record_type
        },
        headers=headers
    )
    return response.json()


if __name__ == "__main__":
    # Example usage
    print("=== List all records ===")
    result = list_records()
    print(json.dumps(result, indent=2))

    print("\n=== List only A records ===")
    result = list_records("A")
    print(json.dumps(result, indent=2))

    print("\n=== Create an A record ===")
    result = create_record("api-test", "A", "10.0.0.1", 300)
    print(json.dumps(result, indent=2))

    print("\n=== Update the record ===")
    result = update_record("api-test", "A", "10.0.0.1", "10.0.0.2", 300)
    print(json.dumps(result, indent=2))

    print("\n=== Delete the record ===")
    result = delete_record("api-test", "A", "10.0.0.2")
    print(json.dumps(result, indent=2))

    print("\n=== DDNS update ===")
    result = ddns_update("home")
    print(json.dumps(result, indent=2))

File: examples/cranedns-api.php
<?php
/**
 * CraneDNS API Client Example - PHP
 *
 * This script demonstrates how to use all CraneDNS API methods.
 *
 * Usage:
 *   1. Set your $apiKey and $baseUrl below
 *   2. Run: php cranedns-api.php
 */

$apiKey = 'your-api-key';
$baseUrl = 'https://namecrane.org/index.php';

/**
 * Make an API request
 *
 * @param string $method HTTP method (GET or POST)
 * @param array $params Query parameters
 * @param array|null $body Request body for POST requests
 * @return array Decoded JSON response
 */
function apiRequest(string $method, array $params = [], ?array $body = null): array
{
    global $apiKey, $baseUrl;

    $params['m'] = 'cranedns';
    $params['action'] = 'api';
    $url = $baseUrl . '?' . http_build_query($params);

    $ch = curl_init();
    curl_setopt_array($ch, [
        CURLOPT_URL => $url,
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_HTTPHEADER => [
            'X-API-Key: ' . $apiKey,
            'Content-Type: application/json'
        ]
    ]);

    if ($body !== null) {
        curl_setopt($ch, CURLOPT_POST, true);
        curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($body));
    }

    $response = curl_exec($ch);
    $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
    curl_close($ch);

    return json_decode($response, true) ?? ['error' => 'Invalid response'];
}

/**
 * List all DNS records
 *
 * Example Response:
 * {
 *     "success": true,
 *     "records": [
 *         {"name": "@", "type": "A", "content": "192.168.1.1", "ttl": 3600},
 *         {"name": "www", "type": "CNAME", "content": "example.com.", "ttl": 3600}
 *     ]
 * }
 *
 * @param string|null $type Optional record type filter
 * @return array
 */
function listRecords(?string $type = null): array
{
    $params = ['method' => 'list'];
    if ($type) {
        $params['type'] = $type;
    }
    return apiRequest('GET', $params);
}

/**
 * Create a new DNS record
 *
 * Example Response:
 * {"success": true, "message": "Record created"}
 *
 * Error Response:
 * {"success": false, "error": "Invalid IPv4 address format"}
 *
 * @param string $name Record name (@ for root)
 * @param string $type Record type (A, AAAA, CNAME, etc.)
 * @param string $content Record value
 * @param int $ttl Time to live in seconds
 * @return array
 */
function createRecord(string $name, string $type, string $content, int $ttl = 3600): array
{
    return apiRequest('POST', ['method' => 'create'], [
        'name' => $name,
        'type' => $type,
        'content' => $content,
        'ttl' => $ttl
    ]);
}

/**
 * Update an existing DNS record
 *
 * Example Response:
 * {"success": true, "message": "Record updated"}
 *
 * Error Response:
 * {"success": false, "error": "name, type, old_content, and content are required"}
 *
 * @param string $name Record name
 * @param string $type Record type
 * @param string $oldContent Current record value
 * @param string $newContent New record value
 * @param int $ttl Time to live in seconds
 * @return array
 */
function updateRecord(string $name, string $type, string $oldContent, string $newContent, int $ttl = 3600): array
{
    return apiRequest('POST', ['method' => 'update'], [
        'name' => $name,
        'type' => $type,
        'old_content' => $oldContent,
        'content' => $newContent,
        'ttl' => $ttl
    ]);
}

/**
 * Delete a DNS record
 *
 * Example Response:
 * {"success": true, "message": "Record deleted"}
 *
 * Error Response:
 * {"success": false, "error": "name, type, and content are required"}
 *
 * @param string $name Record name
 * @param string $type Record type
 * @param string $content Record value
 * @return array
 */
function deleteRecord(string $name, string $type, string $content): array
{
    return apiRequest('POST', ['method' => 'delete'], [
        'name' => $name,
        'type' => $type,
        'content' => $content
    ]);
}

/**
 * DDNS update - update A/AAAA record with client's IP
 *
 * Example Response:
 * {
 *     "success": true,
 *     "message": "DDNS updated",
 *     "name": "home",
 *     "type": "A",
 *     "content": "203.0.113.45"
 * }
 *
 * Error Response:
 * {"success": false, "error": "Could not detect a valid IPv6 address"}
 *
 * @param string $name Record name
 * @param string $type A or AAAA
 * @return array
 */
function ddnsUpdate(string $name, string $type = 'A'): array
{
    return apiRequest('GET', ['method' => 'ddns', 'name' => $name, 'type' => $type]);
}

// Example usage
if (php_sapi_name() === 'cli') {
    echo "=== List all records ===\n";
    print_r(listRecords());

    echo "\n=== List only A records ===\n";
    print_r(listRecords('A'));

    echo "\n=== Create an A record ===\n";
    print_r(createRecord('api-test', 'A', '10.0.0.1', 300));

    echo "\n=== Update the record ===\n";
    print_r(updateRecord('api-test', 'A', '10.0.0.1', '10.0.0.2', 300));

    echo "\n=== Delete the record ===\n";
    print_r(deleteRecord('api-test', 'A', '10.0.0.2'));

    echo "\n=== DDNS update ===\n";
    print_r(ddnsUpdate('home'));
}
